apps/rush-lib/src/logic/buildCache/AmazonS3BuildCacheProvider.ts

```diff
@@ -0,0 +1,206 @@
+
+export interface IAmazonS3BuildCacheProviderOptions {
+  s3Bucket: string;
+  s3Region: string;
+  s3Prefix?: string;
+  isCacheWriteAllowed: boolean;
+}
+
+export class AmazonS3BuildCacheProvider extends CloudBuildCacheProviderBase {
+  private readonly _s3Bucket: string;
+  private readonly _s3Region: string;
+  private readonly _s3Prefix: string | undefined;
+  private readonly _environmentWriteCredential: string | undefined;
+  private readonly _isCacheWriteAllowedByConfiguration: boolean;
+
+  public get isCacheWriteAllowed(): boolean {
+    return this._isCacheWriteAllowedByConfiguration || !!this._environmentWriteCredential;
+  }
+
+  private __s3Client: S3Client | undefined;
+
+  public constructor(options: IAmazonS3BuildCacheProviderOptions) {
+    super();
+    this._s3Bucket = options.s3Bucket;
+    this._s3Region = options.s3Region;
+    this._s3Prefix = options.s3Prefix;
+    this._environmentWriteCredential = EnvironmentConfiguration.buildCacheWriteCredential;
+    this._isCacheWriteAllowedByConfiguration = options.isCacheWriteAllowed;
+  }
+
+  private _deserializeCredentials(credentialString: string | undefined): IAmazonS3Credentials | undefined {
+    if (!credentialString) {
+      return undefined;
+    }
+    const splitIndex: number = credentialString.indexOf(':');
+    if (splitIndex === -1) {
+      return undefined;
+    }
+    return {
+      accessKeyId: credentialString.substring(0, splitIndex),
+      secretAccessKey: credentialString.substring(splitIndex + 1)
+    };
+  }
+
+  private get _credentialCacheId(): string {
+    const cacheIdParts: string[] = ['aws-s3', this._s3Region, this._s3Bucket];
+
+    if (this._isCacheWriteAllowedByConfiguration) {
+      cacheIdParts.push('cacheWriteAllowed');
+    }
+    return cacheIdParts.join('|');
+  }
+
+  private async _getS3ClientAsync(): Promise<S3Client> {
+    if (!this.__s3Client) {
+      let credentials: IAmazonS3Credentials | undefined = this._deserializeCredentials(
+        this._environmentWriteCredential
+      );
+      if (!credentials) {
+        let cacheEntry: ICredentialCacheEntry | undefined;
+        await CredentialCache.usingAsync(
+          {
+            supportEditing: false
+          },
+          (credentialsCache: CredentialCache) => {
+            cacheEntry = credentialsCache.tryGetCacheEntry(this._credentialCacheId);
+          }
+        );
+
+        if (cacheEntry) {
+          const expirationTime: number | undefined = cacheEntry.expires?.getTime();
+          if (expirationTime && expirationTime < Date.now()) {
+            throw new Error(
+              'Cached Amazon S3 credentials have expired. ' +
+                `Update the credentials by running "rush ${RushConstants.updateCloudCredentialsCommandName}".`
+            );
+          } else {
+            credentials = this._deserializeCredentials(cacheEntry?.credential);
+          }
+        } else {
+          try {
+            credentials = await awsCredentialsProvider()();
```

Comment:
Is this an anonymous credential?

Reply:
This goes looking for credentials in:
* Environment variables exposed via process.env
* Shared credentials and config ini files (e.g. ~/.aws/credentials)
* The EC2/ECS Instance Metadata Service, if running in such a container